javaslang/src/main/java/javaslang/Lazy.java

```diff
@@ -135,14 +133,11 @@ private Lazy(Supplier<? extends T> supplier) {
      */
     @Override
     public T get() {
-        // using a local var speeds up the double-check idiom by 25%, see Effective Java, Item 71
-        Supplier<? extends T> tmp = supplier;
-        if (tmp != null) {
+        if ((value == null) && (supplier != null)) {
```

Comment:
Wonderful idea! I have two remarks:

**1) supplier.get() may return null**

Our value has two possible null-states: null as undefined and null as initialized state. If it is initialized with null, the get() calls will be slower than in #1576.

**2) two reads from main memory within the synchronized block**

Because supplier is volatile, all occurrences will be read from main memory, in particular:

* if (supplier != null) { ... }
* value = supplier.get();

The following solution solves 1) and 2) by fusing the ideas of #1577  with those of #1576 and additionally adding an UNDEFINED object:

```java
    private static final Object UNDEFINED = new Object();

    private transient volatile Supplier<? extends T> supplier;

    @SuppressWarnings("unchecked")
    private T value = (T) UNDEFINED;

    public T get() {
        if (value == UNDEFINED && this.supplier != null) {
            synchronized(this) {
                final Supplier<? extends T> supplier = this.supplier;
                if (supplier != null) {
                    value = supplier.get();
                    this.supplier = null;
                }
            }
        }
        return value;
    }
```

I ask myself if the isEvaluated method could now just use value like this:

```java
public boolean isEvaluated() {
    // fast read, memory synchronized with volatile write of supplier
    return value != UNDEFINED;
}
```

instead of

```java
public boolean isEvaluated() {
    // slow read from main memory
    return supplier == null;
}
```